### PR TITLE
Add local agent

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds support for a local agent running the model to generate code on the user's machine.

### CODE REVIEW

File: src/transformers/tools/agents.py [hunk 1]
```diff
@@ -492,3 +494,114 @@ def generate_one(self, prompt, stop):
...
+        ```py
+        import torch
+        from transformers import LocalAgent
+
+        agent = LocalAgent.from_pretrained("bigcode/starcoder", device_map="auto", torch_dtype=torch.bfloat16)
+        agent.run("Draw me a picture of rivers and lakes.")
+        ```
+        """
+        model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path, **kwargs)
+        tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path, **kwargs)
```
Comment (LysandreJik) [inline src/transformers/tools/agents.py:562]: Nice shortcut!

**Review by LysandreJik — APPROVED** [review]
Looks great to me! cc @gante for the criteria :)

File: src/transformers/tools/agents.py [hunk 2]
```diff
@@ -492,3 +494,114 @@ def generate_one(self, prompt, stop):
...
+
+        result = self.tokenizer.decode(outputs[0].tolist()[src_len:])
+        # Inference API returns the stop sequence
+        for stop_seq in stop:
+            if result.endswith(stop_seq):
+                result = result[: -len(stop_seq)]
+        return result
+
+
+class StopSequenceCriteria(StoppingCriteria):
```
Comment (gante) [inline src/transformers/tools/agents.py:588]: Can we:
1. move it to the stopping criteria file? (I've seen a user asking for this)
2. generalize it for the batched input case? (or add a todo, I'd be happy to expand it :D )

Comment (sgugger) [reply src/transformers/tools/agents.py:588]: Maybe you can move it as you add support for the batched case? I don't need it for the agents and it's not an obvious thing to do (that's the reason I didn't put this in the stopping criteria file by the way).

Comment (gante) [reply src/transformers/tools/agents.py:588]: Sure, I can take care of it afterwards 👍

Comment (sgugger) [reply src/transformers/tools/agents.py:588]: Thanks!

**Pull request merged**